File: zero_slip_boundary.py
"""Port of zero_slip_boundary.m."""
from __future__ import annotations

from typing import List, Tuple, Union

import numpy as np

from build_smooth_function import compute_patch_each_layer


def _ramp_zero_cols(h1: int, ramp_choice: str) -> np.ndarray:
    r = (ramp_choice or "").lower()
    if r == "bi_ramp":
        k = 4
    elif r == "qu_ramp_7":
        k = 7
    elif r == "qu_ramp_5":
        k = 5
    else:
        k = 0
    return np.zeros((h1, k), dtype=np.float64)


def zero_slip_boundary(slip_model: np.ndarray,
                       segment_ID: List[int],
                       top_layer_no: Union[int, str],
                       ratio: float,
                       ramp_choice: str,
                       dip_smooth: bool = True) -> Tuple[np.ndarray, np.ndarray]:
    Np = slip_model.shape[0]
    all_fault_id = slip_model[:, 0].astype(int)
    all_patch_id = slip_model[:, 1].astype(int)
    all_layer_id = slip_model[:, 2].astype(int)
    nL = compute_patch_each_layer(slip_model)

    if dip_smooth:
        V = np.zeros(2 * Np, dtype=np.float64)
        d = np.zeros(2 * Np, dtype=np.float64)
    else:
        V = np.zeros(Np, dtype=np.float64)
        d = np.zeros(Np, dtype=np.float64)

    for this_seg in segment_ID:
        mask = all_fault_id == this_seg
        patch_this = all_patch_id[mask]
        layer_this = all_layer_id[mask]
        nL_this = nL[this_seg - 1, :]
        patch_before = int(nL[:this_seg - 1, :].sum())

        if isinstance(top_layer_no, (int, np.integer)):
            patch_top = patch_this[layer_this == int(top_layer_no)]
        elif top_layer_no == "left":
            Nlayer = int(layer_this.max())
            patch_top = np.array(
                [patch_before + 1 + int(nL_this[:jj].sum()) for jj in range(Nlayer)],
                dtype=int,
            )
        elif top_layer_no == "right":
            Nlayer = int(layer_this.max())
            patch_top = np.array(
                [patch_before + int(nL_this[:jj + 1].sum()) for jj in range(Nlayer)],
                dtype=int,
            )
        else:
            raise ValueError(f"Unsupported top_layer_no: {top_layer_no!r}")

        # MATLAB patch indices are 1-based; convert to 0-based for numpy
        strike_indx = patch_top.astype(int) - 1
        dip_indx = strike_indx + Np

        if dip_smooth:
            zero_slip_indx = np.concatenate([strike_indx, dip_indx])
        else:
            zero_slip_indx = strike_indx

        V[zero_slip_indx] = ratio

    W = np.diag(V)
    h1 = W.shape[0]
    rmp = _ramp_zero_cols(h1, ramp_choice)
    W = np.concatenate([W, rmp], axis=1)
    return W, d

File: matlab_code/zero_slip_boundary.m
function [W,d] = zero_slip_boundary(slip_model,segment_ID,top_layer_no,ratio,ramp_choice,varargin)
% add zero-slip constraint at the top surface
% Supplements: add zero-slip constraint at the bottom surface
%           also add zero-slip slip boundary slip at one side of the fault
% options: 'layer_no': add zeros slip for a certain layer (usually top and bottom layer)
%          'left': left side of the fault plane (for patches of the whole layer)
%          'right': right side of the fault plane

Np = size(slip_model,1);
all_fault_id = slip_model(:,1);
all_patch_id = slip_model(:,2);
all_layer_id = slip_model(:,3);
nL = compute_patch_each_layer(slip_model);

d = zeros(2*Np,1);
dip_smooth = true;
if ~isempty(varargin)
    for CC = 1:floor(length(varargin)/2)
        try
            switch lower(varargin{CC*2-1})

                case 'dip_smooth'
                    dip_smooth = varargin{CC*2};
            end
        catch
            error('Unrecognized Keyword');
        end
    end
end
% select this fault segment
NF = length(segment_ID);
 V = zeros(2*Np,1);
    if ~dip_smooth
       V = zeros(Np,1);
       d = zeros(Np,1);
    end

for ii = 1:NF
    this_segment_ID = segment_ID(ii);
    patch_this_segment = all_patch_id(all_fault_id==this_segment_ID);
    layer_this_segment = all_layer_id(all_fault_id==this_segment_ID);
    nL_this_segment = nL(this_segment_ID,:);
    patch_before_this_segment = sum(sum(nL(1:this_segment_ID-1,:)));

    if isnumeric(top_layer_no)
        patch_top_this_segment = patch_this_segment(layer_this_segment == top_layer_no);
    elseif strcmp(top_layer_no,'left')      % start from the first patch of each layer
        Nlayer = max(layer_this_segment);
        patch_top_this_segment = zeros(Nlayer,1);
        for jj = 1:Nlayer
            patch_top_this_segment(jj) = patch_before_this_segment + 1 + sum(nL_this_segment(1:jj-1));
        end
    elseif strcmp(top_layer_no,'right')
        Nlayer = max(layer_this_segment);
        patch_top_this_segment = zeros(Nlayer,1);
        for jj = 1:Nlayer
            patch_top_this_segment(jj) = patch_before_this_segment + sum(nL_this_segment(1:jj));
        end
    else
        error('There is something wrong with the patch options!');
    end

    strike_indx = patch_top_this_segment';
    %     disp(strike_indx);

    dip_indx = strike_indx + Np;
    %     disp(dip_indx);
    if dip_smooth
        zero_slip_indx = [strike_indx,dip_indx];
    else
        zero_slip_indx = [strike_indx];
    end
    V(zero_slip_indx) = ratio;
end
W = diag(V);      % Use diagonal function to speed it up

h1 = size(W,1);
ramp_choice = lower(ramp_choice);
if strcmp(ramp_choice,'bi_ramp')
    rmp = zeros(h1,4);       % bilinear ramp
elseif strcmp(ramp_choice,'qu_ramp_7')
    rmp = zeros(h1,7);
elseif strcmp(ramp_choice,'qu_ramp_5')
    rmp = zeros(h1,5);
else
    rmp = [];
end
W = [W,rmp];

end